sphinx/ext/autodoc/__init__.py

```diff
@@ -389,21 +389,27 @@ def format_signature(self, **kwargs: Any) -> str:
         if self.args is not None:
             # signature given explicitly
             args = "(%s)" % self.args
+            retann = self.retann
         else:
             # try to introspect the signature
             try:
+                retann = None
                 try:
                     args = self.format_args(**kwargs)
                 except TypeError:
                     # retry without arguments for old documenters
                     args = self.format_args()
+
+                if args:
+                    matched = re.match(r'^(\(.*\))\s+->\s+(.*)$', args)
```

Comment:
Won't this fail for a signature like `def evil(text=") -> bad") -> None`?

Reply:
Yes, it would be confused by string. In your example, it works fine because this regexp matches greedily:
```python
>>> sig = '(text=") -> bad") -> None'
>>> re.match(r'^(\(.*\))\s+->\s+(.*)$', sig)
<re.Match object; span=(0, 25), match='(text=") -> bad") -> None'>
>>> _.groups()
('(text=") -> bad")', 'None')
```

But return annotation containing closing-parens and arrow can cheat it.
```python
>>> sig = '(text) -> ") -> None"'
>>> re.match(r'^(\(.*\))\s+->\s+(.*)$', sig)
<re.Match object; span=(0, 21), match='(text) -> ") -> None"'>
>>> _.groups()
('(text) -> ")', 'None"')
```

But I think this will not cause a real problem.